src/hotspot/share/gc/z/c2/zBarrierSetC2.cpp

```diff
@@ -320,6 +320,20 @@ void ZStoreBarrierStubC2::emit_code(MacroAssembler& masm) {
   ZBarrierSet::assembler()->generate_c2_store_barrier_stub(&masm, static_cast<ZStoreBarrierStubC2*>(this));
 }
 
+uint ZBarrierSetC2::estimated_barrier_size(const Node* node) const {
+  uint8_t barrier_data = MemNode::barrier_data(node);
+  assert(barrier_data != 0, "should be a barrier node");
+  uint uncolor_or_color_size = node->is_Load() ? 1 : 2;
+  if ((barrier_data & ZBarrierElided) != 0) {
+    return uncolor_or_color_size;
+  }
+  // A compare and branch corresponds to approximately five Ideal nodes (Cmp,
+  // Bool, If, IfTrue, IfFalse). The runtime call corresponds to approximately
+  // seven more nodes (CallLeaf, control Proj, memory Proj, data Proj, Region,
+  // memory Phi, data Phi).
+  return uncolor_or_color_size + 12;
```

Comment:
I thought the runtime call does not lie inside the loop. Is it necessary to take them into account, too?

Reply:
Conceptually, the runtime call belongs to the loop, even if it is laid out in the cold section of the method. The current unrolling heuristic counts all basic blocks in the loop, regardless of whether they are hot or cold and how they are arranged in the final code. This changeset does the same for consistency.